app/uk/gov/hmrc/selfservicetimetopay/controllers/CalculatorController.scala

```diff
@@ -18,94 +18,302 @@ package uk.gov.hmrc.selfservicetimetopay.controllers
+        } else if (form.get.amount.compare(BigDecimal("32.00")) < 0) {
+          Redirect(routes.SelfServiceTimeToPayController.youNeedToFilePresent())
+        } else {
+          val newAmounts = (amountsDue match {
+            case Some(amounts: CalculatorAmountsDue) => amounts.amountsDue
+            case None => Seq.empty
+          }) :+ form.get
+          Redirect(routes.CalculatorController.amountsDuePresent()).addingToSession(
+            "CalculatorAmountsDue" -> JacksonMapper.writeValueAsString(CalculatorAmountsDue(newAmounts))
+          )
+        }
+      }
+
+      case other => getRedirectionDestination(other)
+
+    })
   }
 
   def paymentTodayPresent: Action[AnyContent] = Action.async { implicit request =>
-    val form = createPaymentTodayForm
-    Future.successful(Ok(payment_today_form.render(form, request)))
+
+    Future.successful(getKeystoreData match {
+
+      case (_, Some(amountsDue:CalculatorAmountsDue), paymentTodayOption: Option[CalculatorPaymentToday], _, _) => {
+        val form = paymentTodayOption match {
+          case Some(paymentToday:CalculatorPaymentToday) => createPaymentTodayForm(amountsDue.total).fill(paymentToday)
+          case None => createPaymentTodayForm(amountsDue.total)
+        }
+        Ok(payment_today_form.render(form, request))
+      }
+
+      case other => getRedirectionDestination(other)
+
+    })
   }
 
+
   def paymentTodaySubmit: Action[AnyContent] = Action.async { implicit request =>
-    val form = createPaymentTodayForm.bindFromRequest()
-    if (form.hasErrors) {
-      Future.successful(Ok(payment_today_form.render(form, request)))
-    } else {
-      Future.successful(Redirect(routes.CalculatorController.calculateInstalmentsPresent()))
-    }
+
+    Future.successful(getKeystoreData match {
+
+      case (_, amountsDue:Some[CalculatorAmountsDue], _, _, _) => {
+        val form = createPaymentTodayForm(amountsDue.get.total).bindFromRequest()
+        if (form.hasErrors) {
+          Ok(payment_today_form.render(form, request))
+        } else {
+          Redirect(routes.CalculatorController.calculateInstalmentsPresent(None)).addingToSession(
+            "CalculatorPaymentToday" -> JacksonMapper.writeValueAsString(form.get)
+          )
+        }
+      }
+
+      case other => getRedirectionDestination(other)
+    })
   }
 
-  def calculateInstalmentsPresent: Action[AnyContent] = Action.async { implicit request =>
-    //@TODO replace with keysotre
-    val calculatorDuration:CalculatorDuration = CalculatorDuration(request.session.get("CalculatorDuration").map(_.toInt)
-      .getOrElse(paymentSchedules.minBy(_.instalments.length).instalments.length))
-    val form = createDurationForm(paymentSchedules.minBy(_.instalments.length).instalments.length,
-      paymentSchedules.maxBy(_.instalments.length).instalments.length).fill(calculatorDuration)
+  def calculateInstalmentsPresent(monthsOption:Option[String]): Action[AnyContent] = Action.async { implicit request =>
```

Comment:
This function is very complex and doesn't have any test coverage. It is factored in a strange way, as pointed out below there is dead code in here. Nested pattern matching is probably not the most elegant way to structure your code, I can understand if it has grown that way "organically" but really you'll need some tests there so you can refactor it later with confidence that you are not breaking things. Same applies to the entire file, lots of logic no test coverage